### ISSUE TITLE
Do we need a `dataset` class just to specify the corresponding `code`

### ISSUE THREAD

**Issue by gogonzo** [insightsengineering/teal.data#139]
We are facing the refactor of `teal.data` to simplify the API and the way how to specify custom DDL app. In the code below, simple specification of `tdata` object is shown where one specifies code via `dataset` function.

```r
data <- teal_data(
  dataset("adsl", adsl, code = "adsl code"),
  dataset("adtte", adtte, code = "adtte code"),
  join_keys
)
```
Technically, above code specification adds a additional flag to the `CodeClass` object, saying "this code element is associated with xyz object". Such `CodeClass` object allows then to pull the code which is related only to a specific dataset (`CodeClass$get_code("adtte")` returns code elements associated with adtte - graph relationships).

In my opinion `TealDataset` class is unnecessary and we are able to specify `teal_data` from simple generic objects, instead of using `dataset` wrapper to create `TealDataset`. 

```r
data <- teal_data(
  adsl, 
  adtte, 
  code = "adsl and adtte code", 
  join_keys
)
```

Above specification has some drawbacks as `code` is specified together for all datasets, which means that returned code will contain pre-processing code for all datasets (even if they are not used in the module).

There is a solution - to alternatively pass code as a code class directly to the `teal_data` instead of indirect through `dataset` function.
Technical efect of the code below would be the same as the first one.

```r
data <- teal_data(
  adsl, 
  adtte,
  code = code_class() |> 
    add_code("adsl code", binding = "adsl") |> 
    add_code("adtte code", binding = "adtte"),
  join_keys
)
```

**Comment by donyunardi** [comment]
I like what you proposed. 
It feels more direct rather having user to construct `dataset` before wrapping it with `teal_data()`.

As for the proposed solution number 3, initially, the `code` part looks a bit odd to me due to the piping activity in the argument. I'm worried that it could be too long if we have many datasets.

I supposed user could define the `code_class` before calling `teal_data`, like so:
```r
code <- code_class |> 
    add_code("adsl code", binding = "adsl") |> 
    add_code("adtte code", binding = "adtte")

data <- teal_data(
    adsl,
    adtte,
    code = code,
    join_keys
)
```

Just an alternative, how about this design?
```r
data <- teal_data(adsl, adrs) |>
    add_code(adsl, code = "adsl <- read_sas('some-adsl-path.sas7bdat')") |>
    add_code(adtte, code = "adtte <- read_sas('some-adtte-path.sas7bdat')")
)
```

I also thought about instantiate an empty teal_data object and build from there but I'm not 100% sure about this one:
```r
data <- teal_data() |>
    add_dataset(adsl, code = "adsl <- read_sas('some-adsl-path.sas7bdat')") |>
    add_dataset(adrs, code = "adrs <- read_sas('some-adsl-path.sas7bdat')")
)
```

**Comment by gogonzo** [comment]
Yes @donyunardi , additional `add_code or `add_dataset`` method for `tdata` holding `CodeClass` would be possible also. I'm happy you agree with the idea of removing `Dataset` class.

**Comment by lcd2yyz** [comment]
> ```r
> data <- teal_data(adsl, adrs) |>
>     add_code(adsl, code = "adsl <- read_sas('some-adsl-path.sas7bdat')") |>
>     add_code(adtte, code = "adtte <- read_sas('some-adtte-path.sas7bdat')")
> )
> ```

I like this, arguments and constructions are clear.
One request is to make sure we can allow both by-dataset processing code (already covered in this proposal), as well as one single file for processing all datasets (bulletin 2 in @gogonzo's proposal). I think it's unavoidable in clinical trial setting that additional processing will involve cross-dataset manipulations, so enabling this will be valuable.
